File: benchmark/corpus/php/sqli_safe_1.php
<?php
// SAFE: PDO prepared statement with bound parameter
function getUser(PDO $pdo, string $id): ?array {
    $stmt = $pdo->prepare("SELECT * FROM users WHERE id = ?");
    $stmt->execute([$id]);
    return $stmt->fetch(PDO::FETCH_ASSOC) ?: null;
}

$pdo  = new PDO("mysql:host=localhost;dbname=mydb", "root", "");
$user = getUser($pdo, $_GET['id'] ?? '');
echo htmlspecialchars($user['username'] ?? '', ENT_QUOTES, 'UTF-8');

File: benchmark/corpus/java/SqliSafe1.java
// SAFE: PreparedStatement with bound parameter prevents SQL injection
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

public class SqliSafe1 {

    public static String findUser(String username) throws Exception {
        Connection conn = DriverManager.getConnection(
                "jdbc:mysql://localhost/app", "root", "");
        PreparedStatement stmt = conn.prepareStatement(
                "SELECT email FROM users WHERE username = ?");
        stmt.setString(1, username);
        ResultSet rs = stmt.executeQuery();
        return rs.next() ? rs.getString("email") : null;
    }

    public static void main(String[] args) throws Exception {
        System.out.println(findUser(args[0]));
    }
}
